tests/functional/test_actions_on_host.py

```diff
@@ -77,19 +85,93 @@ def test_availability_at_host_state(self, cluster: Cluster, provider: Provider):
+        action_in_object_is_present(action_name, host_with_different_services)
+        action_in_object_is_absent(action_name, host_without_component)
+        action_in_object_is_absent(action_name, host_outside_cluster)
+        action_in_object_is_absent(action_name, cluster_with_service)
+        action_in_object_is_absent(action_name, service)
+        run_host_action_and_assert_result(host_with_one_component, action_name)
+        run_host_action_and_assert_result(host_with_two_components, action_name)
+        run_host_action_and_assert_result(host_with_different_services, action_name)
+
+    def test_availability_at_state(self, cluster_with_service: Cluster, provider: Provider):
+        """
+        Test that service host action is available on specify service state
+        """
+        action_name = "action_on_host_state_installed"
+        service = cluster_with_service.service_add(name="Dummy service")
+        host = provider.host_create("host_in_cluster")
+        cluster_with_service.host_add(host)
+        cluster_with_service.hostcomponent_set((host, service.component(name="first")))
+
+        action_in_object_is_absent(action_name, host)
+        run_cluster_action_and_assert_result(cluster_with_service, "switch_cluster_state")
+        action_in_object_is_absent(action_name, host)
+        run_service_action_and_assert_result(service, "switch_service_state")
+        action_in_object_is_present(action_name, host)
+        run_host_action_and_assert_result(host, action_name)
+
+    def test_availability_at_host_state(self, cluster_with_service: Cluster, provider: Provider):
+        """
+        Test that service host action isn't available on specify host state
+        """
+        action_name = "action_on_host_state_installed"
+        service = cluster_with_service.service_add(name="Dummy service")
+        host = provider.host_create("host_in_cluster")
+        cluster_with_service.host_add(host)
+        cluster_with_service.hostcomponent_set((host, service.component(name="first")))
+
+        action_in_object_is_absent(action_name, host)
+        run_host_action_and_assert_result(host, "switch_host_state")
+        action_in_object_is_absent(action_name, host)
+        run_service_action_and_assert_result(service, "switch_service_state")
+        action_in_object_is_present(action_name, host)
+        run_host_action_and_assert_result(host, action_name)
+
+
+ObjTypes = TypeVar("ObjTypes", Cluster, Host, Service)
+
+
+def action_in_object_is_present(action: str, obj: ObjTypes):
     with allure.step(f"Assert that action {action} is present in {_get_object_represent(obj)}"):
         try:
             obj.action(name=action)
         except ObjectNotFound as err:
             raise AssertionError(f"Action {action} not found in object {obj}") from err
 
 
-def action_in_object_is_absent(action: str, obj: Union[Cluster, Host]):
+def action_in_object_is_absent(action: str, obj: ObjTypes):
```

Comment:
Why not `Union[Cluster, Host, Service]`?

Reply:
it's a type alias.